README.md
# CO2System.jl

Port of CO2SYS v2.0.5 into Julia v1.1.

Installation:

```julia
pkg> add https://github.com/mvdh7/CO2System.jl
```

Usage:

```julia
using CO2System

DATA, HEADERS, NICEHEADERS =
    CO2SYS(PAR1,PAR2,PAR1TYPE,PAR2TYPE,SAL,TEMPIN,TEMPOUT,PRESIN,PRESOUT,
    SI,PO4,pHSCALEIN,K1K2CONSTANTS,KSO4CONSTANTS)
```

Inputs can be arrays or scalar or a mixture, just like the MATLAB version. Outputs also correspond as closely as possible to their MATLAB equivalents. For now, please consult the documentation comments within the main function (in [src/CO2System.jl](https://github.com/mvdh7/CO2System.jl/blob/master/src/CO2System.jl)) for full details.

The original MATLAB examples have also been converted into Julia (in [examples/](https://github.com/mvdh7/CO2System.jl/tree/master/examples)). These should work if you have installed CO2System.jl as described above. To run them, download the files and include in the REPL, for example:

```julia
julia> include("CO2SystemExample2.jl")

Here is another example:

What is the pH on the Seawater Scale at 2 degrees, 4000 meters deep of a sample
that has a pH of 7.8 on the Total Scale at 25 degrees at atmospheric pressure?

(Addional info: alk=2400, si=50, po4=2, dissociation constats: Mehrbach Refit)

The answer:
7.99035705953435

See CO2SystemExample2.jl to see what the syntax for this calculation was.

Main.CO2SystemExample2

julia>
```

Note that the first example file requires you to have installed [Plots.jl](https://github.com/JuliaPlots/Plots.jl).

Original README for the MATLAB version follows.

<hr />

**CITATION**

- If you use any CO2SYS related software, please cite the original work by Lewis and Wallace (1998).
- If you use CO2SYS.m, please cite van Heuven et al (2011).
- If you use errors.m or derivnum.m, please cite Orr et al. (2018).

**CO2SYS-MATLAB versions**

- 1.1   (Sept 2011): van Heuven et al. (2011)
- 2.0   (20 Dec 2016): includes uncertainty propagation
- 2.0.1 (11 Oct 2017): supports TEOS-10 standards (conservattive temperature, absolute salinity)
- 2.0.2 (17 Oct 2017): Octave enhancements changed to be MATLAB compatible
- 2.0.3 (4 Jun 2018): examples added as Jupyter notebooks
- 2.0.4 (10 Nov 2018): defaults for standard uncertainties in constants (epK vector and eBt) made consistent with Orr et al. (2018), i.e., final published version
- 2.0.5 (23 Nov 2018): fixed bug in eBt propagation to deriv array (thanks A. Cochon)

**ABOUT CO2SYS**

Here you will find a MATLAB-version of CO2SYS, originally written for
DOS. CO2SYS calculates and returns a detailed state of the carbonate system for
oceanographic water samples, if supplied with sufficient input.  Use the CO2SYS
function as you would use any other MATLAB inline function, i.e.,
a=func(b,c). For much detail about how to use CO2SYS, simply type "help CO2SYS"
in MATLAB.  The help function also works for the two new uncertainty propagation
routines (errors and derivnum).  For details on the internal workings of CO2SYS,
please refer to the original publication (Lewis and Wallace, 1998) available at
http://cdiac.ornl.gov/oceans/co2rprt.html.  Since CO2SYS and the two new
routines each allow input of vectors, with just one call they can process many
samples.  Each sample may have a different salinity, temperature, pH scale,
dissociation constants, etc.

**HISTORY**

The original version for DOS was written by Lewis and Wallace
(1998). That was translated to MATLAB by Denis Pierrot at CIMAS,
University of Miami, Miami, Florida. Then that code was vectorized,
refined, and optimized for computational speed by Steven van Heuven,
University of Groningen, The Netherlands. Although functionality was
added, the output of the CO2SYS function has not changed in form. All
versions of CO2SYS that are available at CDIAC (DOS, Excel, MATLAB)
should produce nearly identical results when supplied with identical
input. Indeed, close agreement between these different versions of
CO2SYS was demonstrated by Orr et al. (2015).  More recently,
CO2SYS-MATLAB has been modified to include uncertainty propagation
(Orr et al., 2018): the main routine CO2SYS.m was altered slightly,
while two new routines were added (errors.m and derivnum.m)

If you discover inconsistencies or have a more general bug report for
CO2SYS.m, please notify S. van Heuven (svheuven at gmail.com), Denis
Pierrot (Denis.Pierrot at noaa.gov), or Alex Kozyr (kozyr at
ornl.gov). For any concerns about the uncertainty propagation routines
(errors.m and derivnum.m), please contact James Orr (james.orr at
lsce.ipsl.fr)

**INSTALLING**

Download the m-files in the src directory (CO2SYS.m, errors.m, and derivnum.m);
you may also wish to download the examples in the examples directory.  Place
these files in a local directory that is in MATLAB's search path, or add the
directory where they are located to MATLAB's search path. The latter can be
done with MATLAB's addpath command, for example

addpath ("my_matlab_directory/my_subdir")

Then run either of the examples in Matlab, or start using the CO2SYS routine
straight away.

**COMPATIBILITY**

Besides their use in MATLAB, the three functions (CO2SYS.m, derivnum.m, and
errors.m) also work well under octave, GNU's MATLAB clone.

**EXAMPLES**

Example MATLAB scripts demonstrating use of CO2SYS can be found in the
examples directory. Using the two new routines is similar, adding only
a few new arguments, e.g., for input uncertainties.  More elaborate
examples are also available in another form in the 'notebooks'
directory. Either click on those files to visualize them (as HTML) or
download those files and use them interactively as jupyter
notebooks. Within MATLAB or octave, you may also use the native help
facility, i.e., by typing "help errors" or "help derivnum" to find out
more.

**REFERENCES**

Lewis, E. and Wallace, D. W. R. (1998) Program Developed for CO2
System Calculations, ORNL/CDIAC-105, Carbon Dioxide Inf.  Anal. Cent.,
Oak Ridge Natl. Lab., Oak Ridge, Tenn., 38 pp.,
https://salish-sea.pnnl.gov/media/ORNL-CDIAC-105.pdf

Orr, J. C., J.-P. Gattuso, and J.-M. Epitalon (2015) Comparison of ten
packages that compute ocean carbonate chemistry, Biogeosciences, 12,
1483–1510, https://doi.org/10.5194/bg-12-1483-2015 .

Orr, J.C., J.-M. Epitalon, A. G. Dickson, and J.-P. Gattuso (2018) Routine
uncertainty propagation for the marine carbon dioxide system, in prep. for
Mar. Chem., in press, https://doi.org/10.1016/j.marchem.2018.10.006.

van Heuven, S., D. Pierrot, J.W.B. Rae, E. Lewis, and D.W.R. Wallace (2011)
MATLAB Program Developed for CO2 System Calculations. ORNL/CDIAC-105b.  Carbon
Dioxide Information Analysis Center, Oak Ridge National Laboratory, U.S.
Department of Energy, Oak Ridge, Tennessee. https://doi.org/10.3334/CDIAC/otg.CO2SYS_MATLAB_v1.1


ChangeLog
ChangeLog

version 2.0.5-jl1, 2019-02-28
-----------------------------------
  - ported main CO2SYS function to Julia, excluding uncertainty propagation parts

version 2.0.5, 2018-11-23
-----------------------------------
	- fixed error in eBt x deriv (array) - Thanks to Anna Conchon for bug report

version 2.0.4, 2018-11-10
-----------------------------------
	- updated README; Orr et al. (2018) published
	- changed defaults of epK and eBt, consistent w/ published version of Orr et al. (2018)

version 2.0.3, 2018-06-04
-----------------------------------
	- changed default uncertainties for epK and eBt for consistency with Orr et al. (2018)
	- new Jupyter notebooks added to demo use of errors, derivnum, CO2SYS routines
	- corrected derivnum.m units of PAR1ref & PAR2ref [H+]
	- fixed eBt error calc in errors.m

version 2.0.2, 2017-10-17
-----------------------------------
	- corrections so that errors.m works when there a vectors in input
	- changed Octave enhancemnts (!, size, resize) to be MATLAB compatible (~, ...)

version 2.0.1, 2017-10-11
-----------------------------------
	- new routines to support use of TEOS-10 standards (Conserrative temperature, Aboslute salinity)
	- added uncertainty in total boron (eBt) as a separate argument in errors.m
	- new and improved examples

version 2.0, 2016-08-15
-----------------------------------
	- errors.m: new routine added to propagate uncertainties
	- derivnum.m: new routine added to compute partial derivatives, needed for error propagation
	- CO2SYS.m: slightly modified routine
		* slight changes to allow error propagation
		* new option to choose K1 & K2 from Waters et al. (2014): fixes inconsistencies with Millero (2010) identified by Orr et al. (2015)

version 1.1, 2016-12-20
-----------------------------------
	- Original version of CO2SYS.m from van Heuven et al. (2011)
	- downloaded on 20 Dec 2016 from CDIAC: http://cdiac.ornl.gov/ftp/co2sys/CO2SYS_calc_MATLAB_v1.1/


REFERENCES

Millero, F. J.: Carbonate constants for estuarine waters, Mar. Freshwater Res., 61, 139–142, doi:10.1071/MF09254, 2010.

Orr, J. C. and Epitalon, J.-M.: Improved routines to model the ocean carbonate system: mocsy 2.0, Geosci. Model Dev., 8, 485–499,
doi:10.5194/gmd–8–485–2015, 2015.

Orr, J. C., Gattuso, J.-P., and Epitalon, J.-M.: Comparison of ten packages that compute ocean carbonate chemistry, Biogeosciences, 12,
35 1483–1510, doi:10.5194/bg–12–1483–2015, 2015.

van Heuven, S., D. Pierrot, J.W.B. Rae, E. Lewis, and D.W.R. Wallace. 2011. MATLAB Program Developed for CO2 System
Calculations. ORNL/CDIAC-105b. Carbon Dioxide Information Analysis Center, Oak Ridge National Laboratory,
U.S. Department of Energy, Oak Ridge, Tennessee. doi: 10.3334/CDIAC/otg.CO2SYS_MATLAB_v1.1

Waters, J., Millero, F., and Woosley, R.: Corrigendum to “The free proton concentration scale for seawater pH”,[MARCHE: 149 (2013)
10 8-22], Mar. Chem., 165, 66–67, 2014.


examples/CO2SYSexample2.jl
module CO2SYSeg2

println(" ")
println("Here is another example:")
println(" ")
println("What is the pH on the Seawater Scale at 2 degrees, 4000 meters deep of a sample")
println("that has a pH of 7.8 on the Total Scale at 25 degrees at atmospheric pressure?")
println(" ")
println("(Addional info: alk=2400, si=50, po4=2, dissociation constats: Mehrbach Refit)")
println(" ")

par1type =   1    # The first parameter supplied is of type "1", which means "alkalinity"
par1    = 2400    # value of the first parameter
par2type =   3    # The first parameter supplied is of type "1", which means "pH"
par2    =  7.8    # value of the second parameter
sal     =   35    # Salinity of the sample
tempin  =   25    # Temperature at input conditions
tempout =    2    # Temperature at output conditions
presin  =    0    # Pressure    at input conditions
presout = 4000    # Pressure    at output conditions
sil     =   50    # Concentration of silicate  in the sample (in umol/kg)
po4     =    2    # Concentration of phosphate in the sample (in umol/kg)
pHscale =    1    # pH scale at which the input pH is reported ("1" means "Total Scale")
k1k2c   =    4    # Choice of H2CO3 and HCO3- dissociation constants K1 and K2 ("4" means "Mehrbach refit")
kso4c   =    1    # Choice of HSO4- dissociation constants KSO4 ("1" means "Dickson")

# Import CO2SYS
include("../src/CO2SYS.jl")
using .CO2SYSjl

# Do the calculation. See CO2SYS's help for syntax and output format
A = CO2SYS(par1,par2,par1type,par2type,sal,tempin,tempout,presin,presout,
    sil,po4,pHscale,k1k2c,kso4c)[1]

println("The answer:") # It is the 38th element of the output of CO2SYS (pHoutSWS)
println(A[38])
println(" ")
println("See CO2SYSexample2.jl to see what the syntax for this calculation was.")
println(" ")

end # module CO2SYSeg2
